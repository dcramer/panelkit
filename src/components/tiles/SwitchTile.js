import PropTypes from "prop-types";

import Tile, { TileProps } from "../Tile";

export default class SwitchTile extends Tile {
  static propTypes = {
    ...TileProps,
    entityId: PropTypes.string.isRequired,
  };

  getWatchedEntityIds() {
    return [this.props.entityId];
  }

  onClick = async () => {
    const { state } = this.getEntity(this.props.entityId);
    await this.callService("switch", state === "on" ? "turn_off" : "turn_on", {
      entity_id: this.props.entityId,
    });
  };

  renderStatus() {}
}
