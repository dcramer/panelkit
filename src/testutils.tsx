import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Entity } from "./types";
import HomeAssistant, { SuggestedChanges } from "./hass";

export const URL = "http://localhost:8123";

export const ACCESS_TOKEN = "example.access.token";

export const DEFAULT_CONFIG = {
  url: URL,
  accessToken: ACCESS_TOKEN,
  tiles: [],
};

export const simulateTouch = (container: any) => {
  userEvent.click(container);
};

export const getTileProps = (entity: Entity) => {
  const hass = new HomeAssistant(DEFAULT_CONFIG);

  hass.callService = jest
    .fn()
    .mockImplementation(
      (
        domain: string,
        service: string,
        serviceData: any,
        suggestedChanges: SuggestedChanges | null = null
      ) => {
        return Promise.resolve();
      }
    );
  hass.sendMessage = jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });

  hass.populateEntityCache([entity]);

  return {
    entityId: entity.entity_id,
    hass,
    cameraList: [],
  };
};

export const renderTile = (
  Component: any,
  entity: Entity,
  props: { [key: string]: any } = {}
) => {
  const tileProps = {
    ...getTileProps(entity),
    ...props,
  };

  const result = render(<Component {...tileProps} />);

  return {
    ...result,
    hass: tileProps.hass,
    container: result.container.firstChild,
  };
};
