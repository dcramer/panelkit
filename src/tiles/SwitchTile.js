import PropTypes from "prop-types";

import Tile, { TileProps } from "./Tile";

export default class SwitchTile extends Tile {
  static propTypes = {
    ...TileProps,
    entityId: PropTypes.string.isRequired,
  };

  getDefaultIcon() {
    const { state } = this.getEntity(this.props.entityId);
    switch (state) {
      case "on":
        return "toggle-switch";
      case "off":
        return "toggle-switch-off";
      default:
        return "toggle-switch";
    }
  }

  onTouch = async () => {
    const { state } = this.getEntity(this.props.entityId);
    await this.callService(
      "switch",
      state === "on" ? "turn_off" : "turn_on",
      {
        entity_id: this.props.entityId,
      },
      {
        [this.props.entityId]: { state: state === "on" ? "off" : "on" },
      }
    );
  };

  renderStatus() {}
}
