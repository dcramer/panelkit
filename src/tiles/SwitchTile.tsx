import Tile, { TileProps } from "./Tile";

type SwitchTileProps = TileProps & {
  entityId: string;
};

export default class SwitchTile extends Tile<SwitchTileProps> {
  getDefaultIcon() {
    const { state } = this.getEntity(this.props.entityId);
    switch (state) {
      case "on":
        return "toggle-switch";
      case "off":
        return "toggle-switch-off";
      default:
        return "toggle-switch";
    }
  }

  onTouch = async () => {
    const { state } = this.getEntity(this.props.entityId);
    await this.callService(
      "switch",
      state === "on" ? "turn_off" : "turn_on",
      {
        entity_id: this.props.entityId,
      },
      {
        [this.props.entityId]: { state: state === "on" ? "off" : "on" },
      }
    );
  };

  renderStatus() {}
}
