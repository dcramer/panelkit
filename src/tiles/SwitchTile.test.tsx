import { waitFor } from "@testing-library/react";

import { renderTile, simulateTouch } from "../testutils";
import SwitchTile from "./SwitchTile";

test("renders", async () => {
  const { container, getByText } = renderTile(SwitchTile, {
    entity_id: "switch.test",
    state: "off",
    attributes: {
      friendly_name: "Test Switch",
    },
  });

  expect(getByText(/Test Switch/i)).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test("turns on on touch event", async () => {
  const { container, hass } = renderTile(SwitchTile, {
    entity_id: "switch.test",
    state: "off",
    attributes: {},
  });

  simulateTouch(container);

  await waitFor(() => {
    expect(hass.callService).toHaveBeenCalledTimes(1);
    expect(hass.callService).toHaveBeenCalledWith(
      "switch",
      "turn_on",
      {
        entity_id: "switch.test",
      },
      {
        "switch.test": { state: "on" },
      }
    );
  });
});

test("turns off on touch event", async () => {
  const { container, hass } = renderTile(SwitchTile, {
    entity_id: "switch.test",
    state: "on",
    attributes: {},
  });

  simulateTouch(container);

  await waitFor(() => {
    expect(hass.callService).toHaveBeenCalledTimes(1);
    expect(hass.callService).toHaveBeenCalledWith(
      "switch",
      "turn_off",
      {
        entity_id: "switch.test",
      },
      {
        "switch.test": { state: "off" },
      }
    );
  });
});
