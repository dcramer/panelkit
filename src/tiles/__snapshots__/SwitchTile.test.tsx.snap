// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="tile state-off touchable"
  data-testid="tile"
  style="cursor: pointer;"
>
  <div
    class="tile-container"
  >
    <div
      class="tile-body"
    >
      <span
        class="tile-icon"
      >
        <svg />
      </span>
    </div>
    <div
      class="tile-title"
    >
      Test Switch
    </div>
  </div>
</div>
`;
